swayfmt/src/utils/language/punctuated.rs

```diff
@@ -78,6 +68,47 @@ where
     }
 }
 
+fn format_generic_pair<T, P>(
+    value_separator_pairs: &[(T, P)],
+    final_value_opt: &Option<Box<T>>,
+    formatter: &mut Formatter,
+) -> Result<FormattedCode, FormatterError>
+where
+    T: Format,
+    P: Format,
+{
+    let mut ps: Vec<String> = value_separator_pairs
+        .iter()
+        .map(|(_, p)| -> Result<FormattedCode, FormatterError> {
+            let mut buf = FormattedCode::new();
+            p.format(&mut buf, formatter)?;
+            Ok(buf)
+        })
+        .collect::<Result<_, _>>()?;
+    let mut ts: Vec<String> = value_separator_pairs
```

Comment:
Is it possible to iterate over `value_separator_pairs` once instead of twice?

Reply:
```rust
    let mut ts = Vec::new();
    let mut ps = Vec::new();

    for (t, p) in value_separator_pairs.iter() {
        let mut t_buf = FormattedCode::new();
        let mut p_buf = FormattedCode::new();
        t.format(&mut t_buf, formatter)?;
        p.format(&mut p_buf, formatter)?;
        ts.push(t_buf);
        ps.push(p_buf);
    }
```

or something similar.

Reply:
It may look nicer to use a `for` loop but it's actually more effective to use `collect` since it doesn't require continual reallocation of memory into a mutable vector. `collect` actually creates a new vec with a known capacity, avoiding those expensive reallocations. 🙂

Reply:
you could pre-allocate the capacity of the vector upfront then.

Reply:
like this

```rust
let len = value_separator_pairs.len();
let mut ts = Vec::with_capacity(len);
let mut ps = Vec::with_capacity(len); 
```